module/os/linux/zfs/zfs_dir.c

```diff
@@ -666,8 +668,9 @@ zfs_rmnode(znode_t *zp)
 			 * Not enough space to delete some xattrs.
 			 * Leave it in the unlinked set.
 			 */
+			zh = zfs_znode_hold_enter(zfsvfs, z_id);
 			zfs_znode_dmu_fini(zp);
-
+			zfs_znode_hold_exit(zfsvfs, zh);
```

Comment:
Moving `zfs_znode_dmu_fini()` under a znode hold means we need to be very careful to avoid performing any `kmem_alloc()`'s in `zfs_znode_dmu_fini()`.  Otherwise, we could deadlock if memory is tight and that allocation enters direct memory reclaim.

It looks to me like there's only one possible allocation under `zfs_znode_dmu_fini()`.  See the call path `zfs_znode_dmu_fini() -> sa_handle_destroy() -> dmu_buf_rele() -> dbuf_rele() -> dbuf_rele_and_unlock() -> arc_buf_freeze() ->  arc_cksum_compute()`.  Removing the allocation entirely and just reserving the space in the structure seems like it should be fine.  These days we're already only adding this structure member when `ZFS_DEBUG` is set anyway.

It'd be worth re-reviewing `zfs_znode_dmu_fini()` again in case I missed any other allocations.  Wrapping these calls in `spl_fstrans_mark()` / `spl_fstrans_unmark()` might be advisable as well in case we missed an allocation, or more likely one gets allocated in the future.

Reply:
`zfs_znode_dmu_fini` under znode hold is pretty much expected though. It's the primary ASSERT, and it always happens for every iput_final for any inode with nonzero link. So I don't really see this as a concern.
Also `zfs_rmnode` should be already wrapped by `spl_fstrans_mark()` in `zpl_evict_inode()`.

Reply:
Good points, yes upon further reflection I believe this should be fine.

Reply:
> Moving `zfs_znode_dmu_fini()` under a znode hold means we need to be very careful to avoid performing any `kmem_alloc()`'s in `zfs_znode_dmu_fini()`. Otherwise, we could deadlock if memory is tight and that allocation enters direct memory reclaim.

`spl_fstrans_mark()` and `spl_fstrans_unmark()` should offer some protection against that. I just opened #14381 to enhance that protection. Anyway, avoiding memory allocations in paths like these certainly is preferable to needing to rely on that protection.